### DISCUSSION TITLE
How does winget choose an MSI over an EXI when the manifest has both?

### DISCUSSION THREAD

**Discussion by mjr4077au** [microsoft/winget-cli#3497] [Q&A]
Take https://github.com/microsoft/winget-pkgs/blob/master/manifests/m/Mozilla/Firefox/115.0.2/Mozilla.Firefox.installer.yaml for example, this manifest offers both MSI and EXE install types for my language and architecture and I was wondering how the MSI is chosen over the EXE?

Example PowerShell code.

```PowerShell
PS C:\Users\User\Desktop> Invoke-RestMethod -UseBasicParsing -Uri https://raw.githubusercontent.com/microsoft/winget-pkgs/master/manifests/m/Mozilla/Firefox/115.0.2/Mozilla.Firefox.installer.yaml | ConvertFrom-Yaml | Select-Object -ExpandProperty Installers | Where-Object {$_.InstallerLocale -eq [System.Globalization.CultureInfo]::CurrentUICulture.Name -and $_.Architecture.Equals('x64')}


InstallerLocale        : en-GB
Architecture           : x64
InstallerType          : wix
Scope                  : machine
InstallerUrl           : https://download-installer.cdn.mozilla.net/pub/firefox/releases/115.0.2/win64/en-GB/Firefox%20Setup%20115.0.2.msi
InstallerSha256        : C04620671E0D5450193F1815172BB0B5337E0DC0DFBF19B85BA58D84385D091F
ProductCode            : {1294A4C5-9977-480F-9497-C0EA1E630130}
AppsAndFeaturesEntries : {@{DisplayName=Mozilla Firefox (x64 en-GB); ProductCode=Mozilla Firefox 115.0.2 (x64 en-GB)}}

InstallerLocale        : en-GB
Architecture           : x64
InstallerType          : exe
Scope                  : machine
InstallerUrl           : https://download-installer.cdn.mozilla.net/pub/firefox/releases/115.0.2/win64/en-GB/Firefox%20Setup%20115.0.2.exe
InstallerSha256        : FBA16C83B68A2EE254564A8879B3737E220A5C28AE3C98E6F3176F3E2DBDE879
InstallerSwitches      : @{Silent=/S; SilentWithProgress=/S}
AppsAndFeaturesEntries : {@{DisplayName=Mozilla Firefox (x64 en-GB); ProductCode=Mozilla Firefox 115.0.2 (x64 en-GB)}}


```
From the source, it appears it purely comes down to the order of the branching in `AppInstaller::Manifest::ConvertToInstallerTypeEnum()` but I'm not 100% sure it's as simple as that?
https://github.com/microsoft/winget-cli/blob/881904c9116bfee91f665cb737f556678eeaf25a/src/AppInstallerCommonCore/Manifest/ManifestCommon.cpp#L120-L167

**Comment by denelon** [comment]
The client now has a "--installer-type" argument that can be used to specify the installer type preferred.

**Reply by mjr4077au** [reply]
Just regarding this, are the accepted values documented? When something incorrect is used, it throws an error but doesn't tell you correct values. I'm assuming the string literals in the source like `wix`, `inno`, etc are what's accepted?
